---
output: pdf_document
params:
    table: "../../../Reports/v0.3 Cortar 03.22/testing/281_MTHFR_P_MTHFR_combined_full.tsv"
    testgenes: "MTHFR"
    sample: "281_MTHFR_P"
    proband: "sj_pct_281_MTHFR_P"
    control_no: "2"
    provider: "AGRF"
    type: "rRNA-depletion on RNA from whole blood"
    matched: "disease-controls run in same flow cell"
    transcripts: c("NM_001330358","NM_001378349","NM_024570")
    refseq: "refseq_introns_exons_hg38.tsv"
    strand: "-"
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library("formattable")
library("data.table")
source("full_gene_figure.R")
```

```{r, echo=FALSE}
#Assign params
sample <- params$sample
proband <- params$proband
control_number <- params$control_no
provider <- params$provider
type <- params$type
batchdate <- params$batchdate
sequencing <- params$sequencing
depth <- params$depth
matched <- params$matched
date <- Sys.Date()
version_number <- "0.3.0"
testgenes <- params$testgenes
transcripts <- params$transcripts
strand <- params$strand
```
# CORTAR v0.3.0
## `r sample`
## Significantly different splicing events (>2SD) compared to `r control_number` controls.
### Details:
- Provider: `r provider`
- RNA-seq: `r type`
- Controls: `r control_number` `r matched`
- Report: `r date`
- Version: v`r version_number`


```{r, echo=FALSE}
table <- fread(params$table)
names(table)[which(names(table) == proband)] <- "probandpct"
```

```{r, echo=FALSE}
exons <- table[SJ_IR == "SJ" ]
introns <- table[SJ_IR == "IR" & normal == "Y"]
```

```{r, echo=FALSE}
table <- rbind(exons, introns)
table$seqnames <- paste0(table$seqnames,":",table$start,"-",table$end)
table$difference <- formattable::percent(table$difference, digits=1)
table$probandpct <- formattable::percent(table$probandpct, digits=1)
table$controlavg <- formattable::percent(table$controlavg, digits=1)
```

```{r, echo=FALSE}
#rnaseqgrapher(table, testgenes, sample, "../../../Reports/v0.3 Cortar 03.22/testing/")
```

```{r, echo=FALSE}
events <- c()

for(row in seq(1,nrow(table))){
    unique <- if(table[row,unique] == "") "" else "*"
    if(table[row,SJ_IR] == "SJ"){
        if(table[row,annotated] == "Y"){
            split <- strsplit(table[row,event]," ")
            redinc <- if(table[row,difference]>0) "increased" else "reduced"
            events[row] <- paste0(redinc, " normal ",split[[1]][1]," ",split[[1]][2],"-",split[[1]][5],unique)
        }else{
            split <- strsplit(table[row,event]," ")
            if(split[[1]][1] != "intron"){
                if(split[[1]][4] == "exon" & split[[1]][1] == "exon"){
                    skips <- seq(as.numeric(split[[1]][2]),as.numeric(split[[1]][5]))
                    skips <- paste(skips[c(-1,-length(skips))],collapse = "-")
                    redinc <- if(table[row,difference]>0) "increased" else "reduced"
                    events[row] <- paste0(redinc," exon ",skips," skipping",unique)
                }else{
                    redinc <- if(table[row,difference]>0) "increased" else "reduced"
                    if (split[[1]][4] == "cryptic") events[row] <- paste0(redinc," ",split[[1]][4]," ",split[[1]][5]," use",unique)
                    if (split[[1]][1] == "cryptic") events[row] <- paste0(redinc," ",split[[1]][1]," ",split[[1]][2]," use",unique)
                }
            }
        }
    }else{
        split <- strsplit(table[row,event]," ")
        redinc <- if(table[row,difference]>0) "increased" else "reduced"
        events[row] <- paste0(redinc," intron ",split[[1]][2]," retention",unique)
    }
}

table$event <- events

table$frame_conserved[which(is.na(table$frame_conserved))] <- ""
table$frame_conserved[which(table$frame_conserved == TRUE)] <- "Yes"
table$frame_conserved[which(table$frame_conserved == FALSE)] <- "No"


```


```{r, echo = FALSE}
table <- table[order(abs(difference),decreasing = T)]

sig_introns <- unique(table[two_sd == TRUE & abs(difference) > 0.05, introns])

no_introns <- length(sig_introns)

sig_introns_no <- as.numeric(sapply(sig_introns, function(x) strsplit(x," ")[[1]][2]))
```
## Overview

In the proband, `r testgenes` has pre-mRNA splicing events which are significantly different to controls (>5% change, >2 standard deviations) in `r no_introns` introns. All pre-mRNA splicing events >2sd occurring in these introns are highlighted below.

Events marked with an asterisk (*) are unique to the proband and any family members included in the analysis, as compared to controls. These events may be present at very low levels in the population.

```{r, echo = FALSE, include=FALSE}
full_gene_figure(testgenes, transcripts, sig_introns_no, strand)
# Close the graphics device
dev.off()
```

![`r testgenes`](full_gene_figure.png)

```{r, results='asis'}
for(i in sig_introns){
    
    cat("## ",i)
    
    introntable <- table[two_sd == TRUE & introns == i]
    insigrow <- colSums(table[two_sd == FALSE & introns == i,.(difference,probandpct,controlavg)])
    insigrow <- as.data.table(t(c("","NS events < 2sd",insigrow,"")))
    insigrow$difference <- formattable::percent(insigrow$difference, digits=1)
    insigrow$probandpct <- formattable::percent(insigrow$probandpct, digits=1)
    insigrow$controlavg <- formattable::percent(insigrow$controlavg, digits=1)
    
    
    events <- introntable[,.(seqnames,event,difference,probandpct,controlavg,frame_conserved)][order(abs(difference), decreasing = T)]
    
    names(events) <- c("splice-junction","event","difference","proband","controls","frame")
    
    events2 <- rbind(events,insigrow, use.names=FALSE)
    
    print(knitr::kable(events2))
    
    cat("\n")

}
```




